pkg/app/app.go

```diff
@@ -48,33 +49,53 @@ func App(ctx context.Context, httpClient *http.Client, configItem *config.Items)
 		return err
 	}
 
+	var wg sync.WaitGroup
+	channel := make(chan error, len(configItem.Bins))
+
+	go func() {
+		wg.Wait()
+		close(channel)
+	}()
+
+	wg.Add(len(configItem.Bins))
+
 	for i := range configItem.Bins {
 		// TODO check configItem.Bins[i].Download == false and create a report function that only is called.
-		err := downloadBin(ctx, client, httpClient, configItem.Bins[i].Owner, configItem.Bins[i].Repo, configItem.Bins[i].Cli, configItem.Bins[i].Tag, configItem.SaveLocation, configItem.Bins[i].Match, configItem.Bins[i].NonGithubURL)
+		//wg.Add(1)
+		go downloadBin(ctx, &wg, channel, client, httpClient, configItem.Bins[i].Owner, configItem.Bins[i].Repo, configItem.Bins[i].Cli, configItem.Bins[i].Tag, configItem.SaveLocation, configItem.Bins[i].Match, configItem.Bins[i].NonGithubURL)
+		fmt.Println("so much stuff")
+
+	}
+
```

Comment:
This is where you need a call to `wg.Wait()`.

Reply:
Assuming that I removed the go func earlier.

Sorry I get a bit confused, On what line do I need the wg.Wait()?
Doesn't it have to be under 
```
	for err := range channel {
		if err != nil {
			close(channel)
			return err
		}
	}
``` 

I tried to have it in a few different spots but when I do it hangs after everything is done.
I guess I need close(channel) as well?
Right now I only close the channel when I hit a error.

Reply:
> Assuming that I removed the go func earlier.
Yes.

The call to `wg.Wait()` should be before looping over the channel (unless you want to do something fancy with canceling the go routines - which I don't recommend.)

Reply:
>  it hangs after everything is done.

If things are hanging, it's _likely_ because you're calling `Add()` more than you're calling `Done()`.

That's one reason to use `wg.Add(1)`, instead of `wg.Add(len(configItem.Bins))`.  The first is just easier to get right.

Reply:
Pushed a new version to make it easier to discuss.
I have performed the other tasks that you pointed out. So I use wg.Add(1).
Also added a simple fmt.Printf("cap: %v, len: %v \n", cap(channel), len(channel))
cap(channel) = 6 just like it should be.

The issue is that when I come to the `for err := range channel {`
The application hangs.

I added a bunch of returns together with the 
`channel <- err` in downloadBin and then the for loop picks up channel and returns it to the main where i perform a os.exit(1) when the rest of the bins is downloaded.
So it works when I add a error to my data.yaml but I assume that's because I do a os.exit(1) and thanks to the for loop helps me to close the application.

But I'm rather confused why the application hangs when all the bins to be downloaded is correctly defined.
The only thing I can come up with is that the channel doesn't get read all the way and never gets emptied or something like that and thus hangs the app.

Reply:
This makes sense because ranging over a channel is a blocking operation - it will wait until there's a message or the channel is closed.  ex. https://gobyexample.com/range-over-channels

Look at using `select` instead - https://gobyexample.com/non-blocking-channel-operations

Reply:
One suggestion.  Because you are only using channels to signal that an error has occurred, consider removing the channels, and just have a boolean variable that indicates that an error occurred.

This would remove a lot of complexity from the code.

Reply:
Yeah with a select it worked as intended :)
Thanks so much for your help.

About the complexity I can only agree, especially since I haven't noticed any significant download improvement compared to before. I think this is due to the github API and all of the IO comes at the same time to my poor laptop.
But the good thing with using the goroutine is that I can download all the bin except the one that failed.

One of the reasons why I write this project is also to be able to try some go concepts so I can learn them so I will probably keep it any way so I have some reference code that I feel I understand for future development.

But when you say that I could have a boolean that indicates if an error occurred how do you mean?
I guess I can't have a goroutine with a return value of a bool?
Do you mean that I should have some global variable that I check for a true in and then trigger some kind of error return?
If I have a goroutine touching a global variable don't I have to screw around with mutex and stuff like that then?

Reply:
> I guess I can't have a goroutine with a return value of a bool?
Correct.  You'll need a shared variable.

> If I have a goroutine touching a global variable don't I have to screw around with mutex and stuff like that then?

It depends on what type of variable you are going to save.  If it's just a boolean (did an error occur?), you don't really need to protect it.  Initialize it to false, and only set it to true if an error occurs.  2 goroutines both setting it to true will work fine.  Just don't set it to false anywhere except initialization.

If you want the error that occurred, then yes, you'll need to use a mutex.

It's just a thought.  I'd only consider tackling that after everything is working.

Reply:
Okay that makes sense.
I will play around with it if nothing else just to learn. Thanks for all your help once again.